akka-javasdk/src/main/scala/akka/javasdk/impl/keyvalueentity/ReflectiveKeyValueEntityRouter.scala

```diff
@@ -5,76 +5,75 @@
 package akka.javasdk.impl.keyvalueentity
 
 import akka.annotation.InternalApi
-import akka.javasdk.JsonSupport
 import akka.javasdk.impl.AnySupport
 import akka.javasdk.impl.CommandHandler
 import akka.javasdk.impl.CommandSerialization
 import akka.javasdk.impl.InvocationContext
 import akka.javasdk.impl.reflection.Reflect
+import akka.javasdk.impl.serialization.JsonSerializer
 import akka.javasdk.keyvalueentity.CommandContext
 import akka.javasdk.keyvalueentity.KeyValueEntity
-import com.google.protobuf.any.{ Any => ScalaPbAny }
+import akka.runtime.sdk.spi.BytesPayload
 
 /**
  * INTERNAL API
  */
 @InternalApi
-private[impl] final class ReflectiveKeyValueEntityRouter[S, E <: KeyValueEntity[S]](
-    override protected val entity: E,
-    commandHandlers: Map[String, CommandHandler])
-    extends KeyValueEntityRouter[S, E](entity) {
+private[impl] class ReflectiveKeyValueEntityRouter[S, KV <: KeyValueEntity[S]](
+    val entity: KV,
+    commandHandlers: Map[String, CommandHandler],
+    serializer: JsonSerializer) {
+
+  val entityStateType: Class[S] = Reflect.keyValueEntityStateType(entity.getClass).asInstanceOf[Class[S]]
 
   private def commandHandlerLookup(commandName: String) =
-    commandHandlers.getOrElse(commandName, throw new RuntimeException(s"no matching method for '$commandName'"))
+    commandHandlers.getOrElse(
+      commandName,
+      throw new HandlerNotFoundException("command", commandName, commandHandlers.keySet))
 
-  override protected def handleCommand(
+  def handleCommand(
       commandName: String,
-      state: S,
-      command: Any,
+      command: BytesPayload,
       commandContext: CommandContext): KeyValueEntity.Effect[_] = {
 
-    _extractAndSetCurrentState(state)
-
     val commandHandler = commandHandlerLookup(commandName)
-    val scalaPbAnyCommand = command.asInstanceOf[ScalaPbAny]
 
-    if (AnySupport.isJson(scalaPbAnyCommand)) {
+    if (serializer.isJson(command)) {
```

Comment:
Just a heads-up

This is only working because the arity 0 payload is using a json content type, which is not really true since is an empty payload. So this test pass for a non-serialized/non-json command.

Later, `deserializeComponentClientCommand` will return a `None` because of empty payload and the arity 0 method will be called. 

I think we will make us a favor if we make it more explicit instead of relying on some conditions to align. 

In the workflow branch I didn't make it much proper either. 

Instead, I think we should:

```scala
if (command == BytesPayload.empty) methodInvoker.invoke(entity)
else methodInvoker.invokeDirectly(entity, deserialize(command)
```

The `InvocationContext` need to be removed on the long run.

Reply:
ok, let's change in separate pr. maybe related to @aludwiko invoker simplifications?

Reply:
I'm not sure if Andrzej is planning to remove the `InvocationContext`, but the reason that we still have it is that we might still receive a payload encoded as json. This is probably some we can fix in the runtime and then we don't need this extra branching here.

Reply:
I'm not removing the `InvocationContext` yet, although I'm not using it in Timers and Consumers and it feels like it's not needed anymore. 

`but the reason that we still have it is that we might still receive a payload encoded as json ...` I don't understand that part, is't it always json?

Reply:
BTW this if in for timed actions looks like this:
```
if ((AnySupport.isJson(updatedContentType) || payload.bytes.isEmpty) && commandHandler.isSingleNameInvoker) {
```

Reply:
I meant to say: encoded as json with protobuff tagging 

This is because the http encoding in the runtime takes the incoming json and wrap it in a proto message and send it down the change. This is only true for the fixed components endpoints. All the other incoming data travels as untagged bytes.